File: app/src/main/java/com/fengniao/news/ui/base/BaseListActivity.java
package com.fengniao.news.ui.base;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.fengniao.news.R;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

public abstract class BaseListActivity<T> extends BaseActivity implements FNAdapter.ViewProvider {
    @BindView(R.id.recycler_view)
    RecyclerView mRecyclerView;
    @BindView(R.id.swipe_refresh)
    SwipeRefreshLayout swipeRefreshLayout;
    protected List<T> mList;
    FNAdapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    int mPage, mCount = 10;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(0);
        initView();
    }

    //刷新控件开关
    public void enableRefresh(boolean status) {
        //禁止下拉刷新
        swipeRefreshLayout.setEnabled(status);
    }

    //是否显示refreshBar
    public void showRefreshBar(boolean isShow) {
        swipeRefreshLayout.setRefreshing(isShow);
    }

    public void setListAdapter() {
        if (getActivity() == null) return;

        if (mAdapter == null) {
            initAdapter();
            mRecyclerView.setAdapter(mAdapter);
        } else
            notifyDataSetChanged();

        if (swipeRefreshLayout.isRefreshing())
            showRefreshBar(false);

        mAdapter.setLoadMoreStatus(false);

    }

    public abstract void loadData();

    @Override
    public void loadMore() {
        mPage++;
        loadData();
    }

    public void initView() {
        mList = new ArrayList<>();
        showRefreshBar(true);
        initAdapter();
        mLayoutManager = onCreateLayoutManager();
        if (mLayoutManager == null)
            mLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setAdapter(mAdapter);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                BaseListActivity.this.onRefresh();
            }
        });
        loadData();
    }


    public void onRefresh() {
        mPage = 0;
        mList.clear();
        loadData();
    }

    public void initAdapter() {
        mAdapter = new FNAdapter(this, mList);
        mAdapter.setViewProvider(this);
    }

    public void enableLoadMore(boolean isLoadMore) {
        mAdapter.enableLoadMore(isLoadMore);
    }

    public RecyclerView.LayoutManager onCreateLayoutManager() {
        return null;
    }

    public void setLayoutManager(RecyclerView.LayoutManager manager) {
        mLayoutManager = manager;
    }

    public void notifyDataSetChanged() {
        mAdapter.notifyDataSetChanged();
    }


    @Override
    public int getItemViewType(int position) {
        return 0;
    }

    @Override
    public int getItemCount() {
        return mList.size();
    }
}

File: app/src/main/java/com/fengniao/news/base/BaseListActivity.kt
package com.fengniao.news.base

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import butterknife.BindView
import com.fengniao.news.R
import java.util.*

abstract class BaseListActivity<T> : BaseActivity(), FNAdapter.ViewProvider {
    @BindView(R.id.recycler_view)
    internal var mRecyclerView: RecyclerView? = null
    @BindView(R.id.swipe_refresh)
    internal var swipeRefreshLayout: SwipeRefreshLayout? = null
    protected lateinit var mList: MutableList<T>
    internal var mAdapter: FNAdapter<*>? = null
    private var mLayoutManager: RecyclerView.LayoutManager? = null
    internal var mPage: Int = 0
    internal var mCount = 10

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(0)
        initView()
    }

    //刷新控件开关
    fun enableRefresh(status: Boolean) {
        //禁止下拉刷新
        swipeRefreshLayout!!.isEnabled = status
    }

    //是否显示refreshBar
    fun showRefreshBar(isShow: Boolean) {
        swipeRefreshLayout!!.isRefreshing = isShow
    }

    fun setListAdapter() {
        if (activity == null) return

        if (mAdapter == null) {
            initAdapter()
            mRecyclerView!!.adapter = mAdapter
        } else
            notifyDataSetChanged()

        if (swipeRefreshLayout!!.isRefreshing)
            showRefreshBar(false)

        mAdapter!!.loadMoreStatus = false

    }

    abstract fun loadData()

    override fun loadMore() {
        mPage++
        loadData()
    }

    fun initView() {
        mList = ArrayList()
        showRefreshBar(true)
        initAdapter()
        mLayoutManager = onCreateLayoutManager()
        if (mLayoutManager == null)
            mLayoutManager = LinearLayoutManager(this)
        mRecyclerView!!.layoutManager = mLayoutManager
        mRecyclerView!!.adapter = mAdapter
        swipeRefreshLayout!!.setOnRefreshListener { this@BaseListActivity.onRefresh() }
        loadData()
    }


    fun onRefresh() {
        mPage = 0
        mList.clear()
        loadData()
    }

    fun initAdapter() {
        mAdapter = FNAdapter(this, mList)
        mAdapter!!.setViewProvider(this)
    }

    fun enableLoadMore(isLoadMore: Boolean) {
        mAdapter!!.enableLoadMore(isLoadMore)
    }

    fun onCreateLayoutManager(): RecyclerView.LayoutManager? {
        return null
    }

    fun setLayoutManager(manager: RecyclerView.LayoutManager) {
        mLayoutManager = manager
    }

    fun notifyDataSetChanged() {
        mAdapter!!.notifyDataSetChanged()
    }


    override fun getItemViewType(position: Int): Int {
        return 0
    }

    override val itemCount: Int
        get() = mList.size
}
